File: src/c/insertionSort.c
#include <stdio.h>

// Função para ordenar um array usando o algoritmo Insertion Sort
void insertionSort(int arr[], int n) {
    // Começa do segundo elemento, pois assumimos que o primeiro já está "ordenado"
    for (int i = 1; i < n; i++) {
        int key = arr[i];  // Número atual a ser inserido na parte ordenada do array
        int j = i - 1;     // Índice do último elemento da parte ordenada

        // Move elementos da parte ordenada que são maiores que 'key' para uma posição à frente
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];  // Move o elemento para a direita
            j--;                  // Avança para o próximo elemento à esquerda
        }

        // Insere o 'key' na posição correta
        arr[j + 1] = key;

        // Exibe o estado do array após cada iteração
        printf("Iteração %d: ", i);
        for (int k = 0; k < n; k++) printf("%d ", arr[k]);
        printf("\n");
    }
}

int main() {
    // Array de exemplo para teste
    int arr[] = {12, 11, 13, 5, 6};

    // Calcula o tamanho do array
    int n = sizeof(arr) / sizeof(arr[0]);

    // Exibe o array antes da ordenação
    printf("Lista antes de ordenar: ");
    for (int i = 0; i < n; i++) printf("%d ", arr[i]);
    printf("\n");

    // Chama a função de ordenação
    insertionSort(arr, n);

    // Exibe o array após a ordenação
    printf("Lista ordenada: ");
    for (int i = 0; i < n; i++) printf("%d ", arr[i]);
    printf("\n");

    return 0;
}

File: src/python/insertionSort.py
def insertion_sort(arr):
    """
    Ordena uma lista de números usando o algoritmo Insertion Sort.
    Args:
        arr (list): A lista de números a ser ordenada.
    """
    # Percorre cada elemento da lista, começando do segundo
    for i in range(1, len(arr)):
        key = arr[i]  # Seleciona o elemento atual para ordenar
        j = i - 1     # Índice do elemento à esquerda da "key"

        # Exibe o estado inicial da iteração
        print(f"Iteração {i}:")
        print(f"  Elemento selecionado: {key}")
        print(f"  Sublista ordenada antes da inserção: {arr[:i]}")

        # Move os elementos maiores que "key" uma posição à direita
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]  # Move o elemento para a direita
            print(f"  {arr[j]} é maior que {key}, movendo para a direita.")
            j -= 1               # Passa para o próximo elemento à esquerda

        # Coloca a "key" na posição correta
        arr[j + 1] = key
        print(f"  Colocando {key} na posição {j + 1}.")
        print(f"  Estado atual da lista: {arr}")
        print()  # Linha em branco para separar as iterações

# Exemplo para testar o algoritmo
arr = [12, 11, 13, 5, 6]  # Lista inicial para teste
print("Lista antes de ordenar:", arr)
insertion_sort(arr)
print("Lista ordenada:", arr)
